2021/OffSeason2021/src/main/cpp/subsystems/Drivetrain.cpp

```diff
@@ -610,69 +556,77 @@ void Drivetrain::MoveWithJoysticksEnd(void)
+    m_slope = (m_dist2 - m_dist1) / (m_targetArea2 - m_targetArea1);
+    m_distOffset = m_dist1 - m_slope * m_targetArea1;
+    frc::SmartDashboard::PutNumber("DTL_Slope", m_slope);
 }
 
-void Drivetrain::MoveWithLimelightExecute(frc::XboxController *throttleJstick)
+void Drivetrain::MoveWithLimelightExecute(double tx, double ta)
 {
     // get turn value - just horizontal offset from target
-    RobotContainer *robotContainer = RobotContainer::GetInstance();
-    double tx = robotContainer->m_vision.GetHorizOffsetDeg();
-    double turnOutput = -m_turnController.Calculate(tx);
-
-    // cap max turn values
-    if (turnOutput < -m_maxTurn)
-        turnOutput = -m_maxTurn;
-    if (turnOutput > m_maxTurn)
-        turnOutput = m_maxTurn;
-
-    frc::SmartDashboard::PutNumber("DTL_TurnOutput", turnOutput);
+    m_tx = tx;
+    double turnOutput = -m_turnController.Calculate(m_tx);
 
     // get throttle value
-    //  get distance based on equation distance = m*targetArea+distOffset
-    double ta = robotContainer->m_vision.GetTargetArea();
+    m_limelightDistance = m_slope * ta - m_distOffset;
 
-    double slope = (m_dist2 - m_dist1) / (m_targetArea2 - m_targetArea1);
-    double distOffset = m_dist1 - slope * m_targetArea1;
-    double limelightDistance = slope * ta - distOffset;
-    double throttleDistance = limelightDistance - m_targetDistance;
-
-    double throttleOutput = m_throttleController.Calculate(throttleDistance);
-    throttleOutput = pow(throttleOutput * cos(turnOutput), m_throttleShape);
-
-    // cap max throttle values
-    if (throttleOutput < -m_maxThrottle)
-        throttleOutput = -m_maxThrottle;
-    if (throttleOutput > m_maxThrottle)
-        throttleOutput = m_maxThrottle;
+    double throttleDistance = m_throttleController.Calculate(m_limelightDistance, m_targetDistance);
+    double throttleOutput = throttleDistance * pow(cos(turnOutput * 180 / wpi::math::pi), m_throttleShape);
 
+    // put turn and throttle outputs on the dashboard
+    frc::SmartDashboard::PutNumber("DTL_TurnOutput", turnOutput);
     frc::SmartDashboard::PutNumber("DTL_ThrottleOutput", throttleOutput);
 
+    // print out inputs and outputs, intermediate values (slope? throttle distance?)
+    spdlog::info(
+        "tx {} ta {} | turn {} throttle {} | limelightDist {} throttleDist {} slope {}",
+        tx,
+        ta,
+        turnOutput,
+        throttleOutput,
+        m_limelightDistance,
+        throttleDistance,
+        m_slope);
+
+    // cap max turn and throttle output
+    turnOutput = std::clamp(turnOutput, -m_maxTurn, m_maxTurn);
+    throttleOutput = std::clamp(throttleOutput, -m_maxThrottle, m_maxThrottle);
+
     if (m_talonValidL1 || m_talonValidR3)
         m_diffDrive.ArcadeDrive(throttleOutput, turnOutput, true);
 }
 
+bool Drivetrain::MoveWithLimelightIsFinished()
+{
+    return (
+        (fabs(m_tx) <= m_angleThreshold)
+        && (fabs(m_limelightDistance - m_targetDistance) <= m_distThreshold));

```

Comment:
Moot issue in this case, but error is calculated as setpoint - measurement. You have them backwards, but since we are taking the absolute value it doesn't matter for the end condition test.